riscv-runtime/src/arith.rs

```diff
@@ -19,52 +29,70 @@ pub(crate) fn u32_to_be(from: &[u32; 8], to: &mut [u8; 32]) {
     }
 }
 
-/// Calculate `a*b + c = hi*2**256 + lo` for 256 bit values (as u8 big-endian arrays).
-/// Returns `(hi, lo)`.
-pub fn affine_256_u8_be(mut a: [u8; 32], mut b: [u8; 32], c: [u8; 32]) -> ([u8; 32], [u8; 32]) {
+/// convert u32 array (arith machine format) to big-endian u16 array
+pub(crate) fn u32x16_to_be(from: &[u32; 16], to: &mut [u8; 64]) {
+    for (i, n) in from.iter().rev().enumerate() {
+        let bytes = n.to_be_bytes();
+        to[i * 4..i * 4 + 4].copy_from_slice(&bytes)
+    }
+}
+
+/// Calculate `a*b + c = d` for 256 bit values (as u8 big-endian arrays).
+/// Returns `d`.
+pub fn affine_256_u8_be(a: [u8; 32], b: [u8; 32], c: [u8; 32]) -> [u8; 64] {
     let mut a1: [u32; 8] = Default::default();
     let mut b1: [u32; 8] = Default::default();
     let mut c1: [u32; 8] = Default::default();
+    let mut res: [u32; 16] = Default::default();
 
     be_to_u32(&a, &mut a1);
     be_to_u32(&b, &mut b1);
     be_to_u32(&c, &mut c1);
 
+    // the Affine256 syscall does x13 = x10 * x11 + x12
+    // where x10, x11, x12 are 256-bit values
+    // and x13 is a 512-bit value.
     unsafe {
-        ecall!(Syscall::Affine256,
-            in("a0") a1.as_mut_ptr(),
-            in("a1") b1.as_mut_ptr(),
-            in("a2") c1.as_ptr());
+        ecall!(
+            Syscall::Affine256,
+            in("a0") a1.as_ptr(),
+            in("a1") b1.as_ptr(),
+            in("a2") c1.as_ptr(),
+            in("a3") res.as_mut_ptr());
     }
 
-    u32_to_be(&a1, &mut a);
-    u32_to_be(&b1, &mut b);
-    (a, b)
+    let mut res_u8: [u8; 64] = [0u8; 64];
+    u32x16_to_be(&res, &mut res_u8);
+    res_u8
 }
 
-/// Calculate `a*b + c = hi*2**256 + lo` for 256 bit values (as u8 little-endian arrays).
-/// Returns `(hi, lo)`.
-pub fn affine_256_u8_le(mut a: [u8; 32], mut b: [u8; 32], c: [u8; 32]) -> ([u8; 32], [u8; 32]) {
+/// Calculate `a*b + c = d` for 256 bit values (as u8 little-endian arrays).
```

Comment:
`d` is 512 bits though right?

Reply:
yes

Reply:
added a comment on the line after this one